process/block/preprocess/intermediateResults.go

```diff
@@ -0,0 +1,162 @@
+) (*intermediateResultsProcessor, error) {
+	if hasher == nil {
+		return nil, process.ErrNilHasher
+	}
+	if marshalizer == nil {
+		return nil, process.ErrNilMarshalizer
+	}
+	if coordinator == nil {
+		return nil, process.ErrNilShardCoordinator
+	}
+	if adrConv == nil {
+		return nil, process.ErrNilAddressConverter
+	}
+
+	irp := &intermediateResultsProcessor{
+		hasher:           hasher,
+		marshalizer:      marshalizer,
+		shardCoordinator: coordinator,
+		adrConv:          adrConv,
+	}
+
+	irp.mutInterResultsForBlock.Lock()
+	irp.interResultsForBlock = make(map[string]*scrInfo, 0)
+	irp.mutInterResultsForBlock.Unlock()
+
+	return nil, nil
+}
+
+// CreateAllScrMiniBlocks returns the cross shard miniblocks for the current round created from the smart contract results
+func (irp *intermediateResultsProcessor) CreateAllInterMiniBlocks() []*block.MiniBlock {
+	miniBlocks := make([]*block.MiniBlock, irp.shardCoordinator.NumberOfShards())
+
+	irp.mutInterResultsForBlock.Lock()
+	defer irp.mutInterResultsForBlock.Unlock()
+
+	for key, value := range irp.interResultsForBlock {
+		if value.receiverShardID != irp.shardCoordinator.SelfId() {
+			miniBlocks[value.receiverShardID].TxHashes = append(miniBlocks[value.receiverShardID].TxHashes, []byte(key))
+		}
+	}
+
+	for i := 0; i < len(miniBlocks); i++ {
+		if len(miniBlocks[i].TxHashes) > 0 {
+			miniBlocks[i].SenderShardID = irp.shardCoordinator.SelfId()
+			miniBlocks[i].ReceiverShardID = uint32(i)
+			miniBlocks[i].Type = block.SmartContractResultBlock
+		}
+	}
+
+	irp.interResultsForBlock = make(map[string]*scrInfo, 0)
+
+	return miniBlocks
+}
+
+// VerifyScrMiniBlocks validity verifies if the smart contract results added to the block are valid
+func (irp *intermediateResultsProcessor) VerifyInterMiniBlocks(body block.Body) error {
+	scrMbs := irp.CreateAllInterMiniBlocks()
+
+	for i := 0; i < len(body); i++ {
+		mb := body[i]
+		if mb.Type != block.SmartContractResultBlock {
+			continue
+		}
+		if mb.ReceiverShardID == irp.shardCoordinator.SelfId() {
+			continue
+		}
+
+		createdScrMb := scrMbs[mb.ReceiverShardID]
+		createdHash, err := core.CalculateHash(irp.marshalizer, irp.hasher, createdScrMb)
+		if err != nil {
+			return err
+		}
+
+		receivedHash, err := core.CalculateHash(irp.marshalizer, irp.hasher, mb)
+		if err != nil {
+			return err
+		}
+
+		if !bytes.Equal(createdHash, receivedHash) {
+			return process.ErrMiniBlockHashMismatch
+		}
+	}
+
+	return nil
+}
+
+// AddIntermediateTransactions adds smart contract results from smart contract processing for cross-shard calls
+func (irp *intermediateResultsProcessor) AddIntermediateTransactions(txs []data.TransactionHandler) error {
+	irp.mutInterResultsForBlock.Lock()
+	defer irp.mutInterResultsForBlock.Unlock()
+
+	for i := 0; i < len(txs); i++ {
+		addScr, ok := txs[i].(*smartContractResult.SmartContractResult)
+		if !ok {
+			return process.ErrWrongTypeAssertion
+		}
+
+		scrMrs, err := irp.marshalizer.Marshal(addScr)
```

Comment:
You can make use of the core.CalculateHash

Reply:
done